microscopic-monks/primal/engine/world.py

```diff
@@ -0,0 +1,229 @@
+            self.image = 'i.png'
+            self.type = 3
+        elif self.sample > 0.5:
+            self.image = 'l.png'
+            self.type = 2
+        elif self.sample > 0.25:
+            self.image = 's.png'
+            self.type = 1
+        else:
+            self.image = 'w.png'
+            self.type = 0
+
+        self.generate_terrain()
+
+    def generate_terrain(self):
+        if self.type == 0:
+            return
+
+        while random.randint(0, 1) == 1:
+            s = random.randint(100, 150)
+            angle = 0
+            sprite = 'r.png'
+            rock = Feature(sprite, Chunk.get_random_position(self.pos, s), .0,
+                           (s, s), angle, 'rock', True)
+            rock.feature.set_angle(random.randint(0, 360))
+            self.chunk_features.add(rock)
+
+        if self.type == 1:
+            while random.randint(0, 1) != 1:
+                s = 65
+                angle = random.randint(0, 359)
+                type = 'cacti'
+                sprite = 'Cacti.png'
+
+                self.chunk_features.add(Feature(sprite, Chunk.get_random_position(self.pos, s),
+                                                1.0, (s, s), angle, type, True))
+        elif self.type == 2:
+            bushes = ['bushBB.png', 'bushBO.png', 'bushBP.png', 'bushBR.png', 'bushBY.png']
+            while random.randint(0, 1) != 1:
+                s = 65
+                angle = random.randint(0, 359)
+                sprite = random.choice(bushes)
+                type = sprite.replace('.png', '')
+                self.chunk_features.add(Feature(sprite, Chunk.get_random_position(self.pos, s),
+                                                1.0, (s, s), angle, type))
+
+            while random.randint(0, 3) != 1:
+                s = random.randint(150, 280)
+                angle = random.randint(0, 359)
+                sprite = 'topOfTree.png'
+
+                self.chunk_features.add(Feature(sprite, Chunk.get_random_position(self.pos, s),
+                                                2.0, (s, s), angle, 'tree'))
+
+    def draw(self, terrain: Sprite):
+        terrain.set_position(self.pos)
+        terrain.set_source(self.image)
+
+    def sort_key(self, f: Feature):
+        return f.get_z()
+
+    def draw_features(self, group: InstructionGroup, top_group: InstructionGroup):
+        list = sorted(self.chunk_features, key=self.sort_key)
+
+        for feature in list:
+            if feature.get_z() >= 2.0:
+                feature.draw(top_group)
+            else:
+                feature.draw(group)
+
+    def get_features(self) -> Set[Feature]:
+        return self.chunk_features
+
+    def remove_feature(self, feature: Feature):
+        self.chunk_features.discard(feature)
+
+    @staticmethod
+    def get_random_position(pos, size):
+        return (
+            pos[0] + random.randint(0, Chunk.SIZE - size),
+            pos[1] + random.randint(0, Chunk.SIZE - size),
+        )
```

Comment:
This module is really well written, it uses many pythonic techniques and is easy to read. There are some sparse annotations but no docstrings.